src/plots/cartesian/set_convert.js

```diff
@@ -184,10 +184,13 @@ module.exports = function setConvert(ax) {
             // that aren't in the first etc.
             // TODO: sorting options - do the sorting
             // progressively here as we insert?
-            if(ax._categories.indexOf(v)===-1) ax._categories.push(v);
 
-            var c = ax._categories.indexOf(v);
-            return c===-1 ? constants.BADNUM : c;
+            if(v !== null && v !== undefined && ax._categories.indexOf(v) === -1){
+                ax._categories.push(v);
+                return ax._categories.length - 1;
```

Comment:
Oh. I don't think this is right.

Because different traces can share categories, so the index of the incoming category here won't necessarily be the `length - 1`.